core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala

```diff
@@ -0,0 +1,428 @@
+    try {
+      var correlationId: Int = 1
+      for (apiKey <- RequestQuotaTest.ClientActions) {
+        correlationId += 1
+        val response = requestResponse(socket, unthrottledClientId, correlationId, requestBuilder(apiKey))
+      }
+    } finally {
+      socket.close()
+    }
+  }
+
+  @Test
+  def testExemptRequestTime() {
+
+    for (apiKey <- RequestQuotaTest.ClusterActions) {
+      submitTest(apiKey, () => checkExemptRequestMetric(apiKey))
+    }
+    waitAndCheckResults()
+  }
+
+  @Test
+  def testUnauthorizedThrottle() {
+
+    RequestQuotaTest.principal = RequestQuotaTest.UnauthorizedPrincipal
+
+    for (apiKey <- ApiKeys.values) {
+      submitTest(apiKey, () => checkUnauthorizedRequestThrottle(apiKey))
+    }
+    waitAndCheckResults()
+  }
+
+  private def throttleTimeMetricValue(clientId: String): Double = {
+    val metricName = leaderNode.metrics.metricName("throttle-time",
+                                  QuotaType.Request.toString,
+                                  "",
+                                  "user", "",
+                                  "client-id", clientId)
+    val sensor = leaderNode.quotaManagers.request.getOrCreateQuotaSensors("ANONYMOUS", clientId).throttleTimeSensor
+    metricValue(leaderNode.metrics.metrics.get(metricName), sensor)
+  }
+
+  private def requestTimeMetricValue(clientId: String): Double = {
+    val metricName = leaderNode.metrics.metricName("request-time",
+                                  QuotaType.Request.toString,
+                                  "",
+                                  "user", "",
+                                  "client-id", clientId)
+    val sensor = leaderNode.quotaManagers.request.getOrCreateQuotaSensors("ANONYMOUS", clientId).quotaSensor
+    metricValue(leaderNode.metrics.metrics.get(metricName), sensor)
+  }
+
+  private def exemptRequestMetricValue: Double = {
+    val metricName = leaderNode.metrics.metricName("exempt-request-time", QuotaType.Request.toString, "")
+    metricValue(leaderNode.metrics.metrics.get(metricName), leaderNode.quotaManagers.request.exemptSensor)
+  }
+
+  private def metricValue(metric: KafkaMetric, sensor: Sensor): Double = {
+    sensor.synchronized {
+      if (metric == null) -1.0 else metric.value
+    }
+  }
+
+  private def requestBuilder(apiKey: ApiKeys): AbstractRequest.Builder[_ <: AbstractRequest] = {
+    apiKey match {
+        case ApiKeys.PRODUCE =>
+          new requests.ProduceRequest.Builder(RecordBatch.CURRENT_MAGIC_VALUE, 1, 5000,
+            collection.mutable.Map(tp -> MemoryRecords.withRecords(CompressionType.NONE, new SimpleRecord("test".getBytes))).asJava)
+
+        case ApiKeys.FETCH =>
+          val partitionMap = new LinkedHashMap[TopicPartition, requests.FetchRequest.PartitionData]
+          partitionMap.put(tp, new requests.FetchRequest.PartitionData(0, 0, 100))
+          requests.FetchRequest.Builder.forReplica(ApiKeys.FETCH.latestVersion, 5000, 100, Int.MaxValue, partitionMap)
```

Comment:
Why is the replicaId 5000? That indicates it's from a follower. Also, I am wondering if 100 maxWait is enough to trigger throttling.